src/test/scala/advancepassengerinfo/importer/processor/DqFileProcessorImplTest.scala

```diff
@@ -253,27 +267,135 @@ class DqFileProcessorTest extends TestKit(ActorSystem("MySpec"))
-        ("manifest2.json", Success(manifest)),
-      )
-      val processor = DqFileProcessorImpl(singleZipMockProvider(manifests), mockPersistence)
-      val dqApiFeed: DqApiFeedImpl = DqApiFeedImpl(
-        MockFileNames(List(List("1.zip", "2.zip"), List("3.zip", "4.zip"))),
+    val manifestsWithFileName: Map[String, List[List[Try[Seq[(String, Try[VoyageManifest])]]]]] = Map(
+      "1.zip" -> List(List(Success(Seq(("manifest1.json", Success(manifest)))))),
+      "2.zip" -> List(List(Success(Seq(("manifest2.json", Success(manifest)))))),
+      "3.zip" -> List(List(Success(Seq(("manifest3.json", Success(manifest)))))),
+      "4.zip" -> List(List(Success(Seq(("manifest4.json", Success(manifest))))))
+    )
+
+    def getDqApiFeedInstance(mockPersistence: MockPersistence): DqApiFeedImpl = {
+      val processor = DqFileProcessorImpl(multipleZipMockProvider(manifestsWithFileName), mockPersistence)
+      DqApiFeedImpl(
+        MockFileNames(batchedFileNames),
         processor,
-        100.millis,
+        200.millis,
         MockStatsDCollector,
         LastCheckedState()
       )
+    }
 
-      dqApiFeed.processFilesAfter("1.zip").runWith(Sink.seq)
+    def afterRecoveryExpectedMsg(probe: TestProbe): ZipFileCall = {
       probe.expectMsg(ManifestCall("manifest2.json", manifest))
       probe.expectMsg(JsonFileCall("2.zip", "manifest2.json", successful = true, dateIsSuspicious = false))
       probe.expectMsg(ZipFileCall("2.zip", successful = true))
+      probe.expectMsg(ManifestCall("manifest3.json", manifest))
+      probe.expectMsg(JsonFileCall("3.zip", "manifest3.json", successful = true, dateIsSuspicious = false))
+      probe.expectMsg(ZipFileCall("3.zip", successful = true))
+      probe.expectMsg(ManifestCall("manifest4.json", manifest))
+      probe.expectMsg(JsonFileCall("4.zip", "manifest4.json", successful = true, dateIsSuspicious = false))
+      probe.expectMsg(ZipFileCall("4.zip", successful = true))
+    }
+
+    "process and persist zip files successful with manifest json" in {
+      val probe: TestProbe = TestProbe("probe")
+      val mockPersistence = new MockPersistence(probe.ref)
+      val dqApiFeed = getDqApiFeedInstance(mockPersistence)
+
+      dqApiFeed.processFilesAfter("_").runWith(Sink.seq)
+      probe.expectMsg(ManifestCall("manifest1.json", manifest))
+      probe.expectMsg(JsonFileCall("1.zip", "manifest1.json", successful = true, dateIsSuspicious = false))
+      probe.expectMsg(ZipFileCall("1.zip", successful = true))
+      afterRecoveryExpectedMsg(probe)
     }
+
+    "process zip files and recover while exception in jsonHasBeenProcessed" in {
+      val probe: TestProbe = TestProbe("probe")
+      var isFirstCall: Boolean = true
+      val mockPersistence = new MockPersistence(probe.ref) {
+        override def jsonHasBeenProcessed(zipFileName: String, jsonFileName: String): Future[Boolean] =
+          if (isFirstCall) {
+            isFirstCall = !isFirstCall
```

Comment:
This should say `isFirstCall = false` to avoid confusion else it looks like it toggles true and false